### PR TITLE
Add CSRF protection 

### PR DESCRIPTION
**Pull request by jseppi**
- Add express [`csurf` middleware](https://github.com/expressjs/csurf) and protect API routes:
  - all `sites` routes
  - `builds` `POST` route
- Adjust client requests to include csrf token in `x-csrf-token` header
- Modify test `support/session.js` so we can create both authenticated and unauthenticated sessions
- Lots of linting fixes

To Do:
- [x] determine if any other routes need csrf protection enabled
- [x] make sure new user can login for the first time -- possible bug here
- [x] remove `x-csrf-token` from front-end requests to GitHub. Their cors settings do not allow it.
- [x] test that requests fail with a specific csrf error if the token is invalid

### CODE REVIEW

**Comment by jseppi** [comment]
Ok, @jmhooper - this is ready for review. It is a pretty big PR because of all the tests that had to be adjusted, but the "core" stuff of it is in `api/controllers/main.js`, `api/routers/build.js`, and `api/routers/site.js`.

File: frontend/components/site/siteSettings.js [hunk 1]
```diff
@@ -140,7 +140,7 @@ export class SiteSettings extends React.Component {
               <div className="well">
                 <h3 className="well-heading">Demo Site</h3>
                 <p className="well-text">
-                  Setup a branch to be deployed to a demo url.
+                  Setup a branch to be deployed to a demo URL.
```
Comment (jseppi) [inline frontend/components/site/siteSettings.js]: minor unrelated fix -- consistent capitalization with other instances of "URL"

File: test/api/support/session.js [hunk 2]
```diff
@@ -30,6 +59,9 @@ const session = (user) => {
...
       return `${config.session.key}=s%3A${signedSessionKey}`;
     });
-};
+}
 
-module.exports = session;
+module.exports = {
+  authenticatedSession,
+  unauthenticatedSession,
+};
```
Comment (jseppi) [inline test/api/support/session.js]: modified this file to export two different session functions for testing: an authenticated one, and an unauthenticated one. The unauthenticated one is so that we can get through the csrf protection middleware in order to actually test that authenticate is properly working.

File: test/api/support/validateAgainstJSONSchema.js [hunk 3]
```diff
@@ -1,25 +1,27 @@
...
+  const statusCodeInt = parseInt(statusCode, 10);
 
-  const responseSchema = schema.paths[path][action].responses[statusCode].schema
-  const result = validate(response,responseSchema)
+  const responseSchema = schema.paths[pathLower][actionLower].responses[statusCodeInt].schema;
+  const result = validate(response, responseSchema);
 
   if (result.errors.length) {
-    console.error(result.errors)
-    throw new Error(`Failed to validate against definition: ${definition}`)
```
Comment (jseppi) [inline test/api/support/validateAgainstJSONSchema.js]: Fixed a minor bug here -- `definition` is not defined anywhere so the error message wasn't very helpful. Since I fixed that bug, I also had to lint the file.

File: views/app.njk [hunk 4]
```diff
@@ -25,5 +25,8 @@
 {% block main_content %}
   <div id="js-app"></div>
 
+  <script>
+    window.CSRF_TOKEN = "{{ csrfToken }}";
```
Comment (jseppi) [inline views/app.njk]: expose the CSRF token as a global var so that it can be grabbed by our `fetch` implementation

File: frontend/util/fetch.js [hunk 5]
```diff
@@ -1,18 +1,22 @@
...
 const credentials = 'same-origin';
+const csrfToken = typeof window !== 'undefined' ?
+  window.CSRF_TOKEN : global.CSRF_TOKEN;
+
 const defaultHeaders = {
-  'Accept': 'application/json',
-  'Content-Type': 'application/json'
+  accept: 'application/json',
+  'content-type': 'application/json',
+  'x-csrf-token': csrfToken,
```
Comment (jseppi) [inline frontend/util/fetch.js]: include the `x-csrf-token` header in all fetch requests from our front end code

File: frontend/util/fetch.js [hunk 6]
```diff
@@ -1,18 +1,22 @@
+/* global window:true fetch:true */
+
 import 'whatwg-fetch';
 
 const defaultMethod = 'GET';
 const credentials = 'same-origin';
+const csrfToken = typeof window !== 'undefined' ?
+  window.CSRF_TOKEN : global.CSRF_TOKEN;
```
Comment (jseppi) [inline frontend/util/fetch.js]: grab the globally exposed `CSRF_TOKEN` variable

**Comment by jseppi** [comment]
Eh, changed back to `WIP` to see if I can improve the tests a bit

File: frontend/util/federalistApi.js [hunk 7]
```diff
@@ -1,14 +1,25 @@
...
 import alertActions from '../actions/alertActions';
 
 export const API = '/v0';
 
+const csrfToken = typeof window !== 'undefined' ?
+  window.CSRF_TOKEN : global.CSRF_TOKEN;
+
+const defaultHeaders = {
+  'x-csrf-token': csrfToken,
+};
```
Comment (jseppi) [inline frontend/util/federalistApi.js]: always include the csrf token header on requests to the Federalist API

File: test/api/requests/build.test.js [hunk 8]
```diff
@@ -32,30 +34,66 @@ describe('Build API', () => {
...
+            .expect(403)
+        )
+        .then((response) => {
+          validateAgainstJSONSchema('POST', '/build', 403, response.body);
+          done();
+        })
+        .catch(done);
     });
 
+    it('should require a valid csrf token', (done) => {
```
Comment (jmhooper) [inline test/api/requests/build.test.js]: 👍
